crates/ruma-common/src/events/audio.rs

```diff
@@ -0,0 +1,180 @@
+//! Types for extensible audio message events ([MSC3246]).
+//!
+//! [MSC3246]: https://github.com/matrix-org/matrix-spec-proposals/pull/3246
+
+use std::{convert::TryFrom, time::Duration};
+
+use js_int::UInt;
+use ruma_macros::EventContent;
+use serde::{Deserialize, Serialize};
+
+mod amplitude_serde;
+mod waveform_serde;
+
+use waveform_serde::WaveformSerDeHelper;
+
+use super::{file::FileContent, message::MessageContent, room::message::Relation};
+
+/// The payload for an extensible audio message.
+#[derive(Clone, Debug, Serialize, Deserialize, EventContent)]
+#[cfg_attr(not(feature = "unstable-exhaustive-types"), non_exhaustive)]
+#[ruma_event(type = "m.audio", kind = MessageLike)]
+pub struct AudioEventContent {
+    /// The text representation of the message.
+    #[serde(flatten)]
+    pub message: MessageContent,
+
+    /// The file content of the message.
+    #[serde(rename = "org.matrix.msc1767.file")]
+    pub file: FileContent,
+
+    /// The audio content of the message.
+    #[serde(rename = "org.matrix.msc1767.audio")]
+    pub audio: AudioContent,
+
+    /// Information about related messages.
+    #[serde(flatten, skip_serializing_if = "Option::is_none")]
+    pub relates_to: Option<Relation>,
+}
+
+impl AudioEventContent {
+    /// Creates a new `AudioEventContent` with the given plain text message and file.
+    pub fn plain(message: impl Into<String>, file: FileContent) -> Self {
+        Self {
+            message: MessageContent::plain(message),
+            file,
+            audio: Default::default(),
+            relates_to: None,
+        }
+    }
+
+    /// Creates a new `AudioEventContent` with the given message and file.
+    pub fn with_message(message: MessageContent, file: FileContent) -> Self {
+        Self { message, file, audio: Default::default(), relates_to: None }
+    }
+}
+
+/// Audio content.
+#[derive(Clone, Debug, Default, Serialize, Deserialize)]
+#[cfg_attr(not(feature = "unstable-exhaustive-types"), non_exhaustive)]
+pub struct AudioContent {
+    /// The duration of the video in milliseconds.
+    #[serde(
+        with = "ruma_common::serde::duration::opt_ms",
+        default,
+        skip_serializing_if = "Option::is_none"
+    )]
+    pub duration: Option<Duration>,
+
+    /// The waveform representation of the audio content.
+    #[serde(default, skip_serializing_if = "Waveform::is_empty")]
+    pub waveform: Waveform,
```

Comment:
I think it's weird to have this type with an invariant of `len == 0 || len >= 30 && len <= 120`. I'd prefer if this was an `Option`.